lib/Sema/CSFix.cpp

```diff
@@ -170,18 +170,166 @@ MarkExplicitlyEscaping::create(ConstraintSystem &cs, Type lhs, Type rhs,
   return new (cs.getAllocator()) MarkExplicitlyEscaping(cs, lhs, rhs, locator);
 }
 
-bool RelabelArguments::diagnose(const Solution &solution, bool asNote) const {
-  LabelingFailure failure(solution, getLocator(), getLabels());
-  return failure.diagnose(asNote);
+// Find RelabelArgument and MoveOutOfOrder fixes.
+// Result is sorted by priority for emitting aggregated diagnostics.
+static void
+findLabelingFixes(const Solution &solution, Expr *applyAnchor,
+                  SmallVectorImpl<const ConstraintFix *> &foundFixes) {
+  auto checkLocatorPrefix = [&](ConstraintLocator *locator) -> bool {
+    if (locator->getAnchor() != applyAnchor)
+      return false;
+
+    if (locator->getPath().size() < 1)
+      return false;
+
+    for (auto element : locator->getPath()) {
+      if (element.getKind() == ConstraintLocator::ApplyArgument) {
+        return true;
+      }
+    }
+    return false;
+  };
+
+  for (auto *checkFix : solution.Fixes) {
+    if (!checkLocatorPrefix(checkFix->getLocator()))
+      continue;
+
+    if (auto *fix = checkFix->getAs<RelabelArgument>()) {
+      foundFixes.push_back(fix);
+      continue;
+    }
+    if (auto *fix = checkFix->getAs<MoveOutOfOrderArgument>()) {
+      foundFixes.push_back(fix);
+      continue;
+    }
+  }
+
+  auto typePriority = [](const ConstraintFix *csFix) -> int {
+    if (auto *fix = csFix->getAs<RelabelArgument>())
+      return 2;
+    if (auto *fix = csFix->getAs<MoveOutOfOrderArgument>())
+      return 1;
+    return 0;
+  };
+
+  auto getArgIdx = [](ConstraintLocator *loc) -> unsigned {
+    if (loc->getPath().empty())
+      return 0;
+    auto element = loc->getPath().back();
+    if (auto elem = element.getAs<LocatorPathElt::ArgumentLabel>()) {
+      return elem->getArgIdx();
+    }
+    return 0;
+  };
+
+  auto compare = [&](const ConstraintFix *a, const ConstraintFix *b) -> bool {
+    auto ap = typePriority(a);
+    auto bp = typePriority(b);
+    if (ap != bp)
+      return ap < bp;
+    auto aa = getArgIdx(a->getLocator());
+    auto ba = getArgIdx(b->getLocator());
+    return aa < ba;
+  };
+
+  llvm::sort(foundFixes, compare);
 }
 
-RelabelArguments *
-RelabelArguments::create(ConstraintSystem &cs,
-                         llvm::ArrayRef<Identifier> correctLabels,
-                         ConstraintLocator *locator) {
-  unsigned size = totalSizeToAlloc<Identifier>(correctLabels.size());
-  void *mem = cs.getAllocator().Allocate(size, alignof(RelabelArguments));
-  return new (mem) RelabelArguments(cs, correctLabels, locator);
+static bool diagnoseLabelError(const ConstraintFix *thisFix,
+                               ArrayRef<const ConstraintFix *> fixes,
+                               const Solution &solution,
+                               ConstraintLocator *locator,
+                               ArrayRef<ParamBinding> bindings, bool asNote) {
```

Comment:
It seems like this logic belongs in the diagnostic itself, also if this is intended to diagnose labeling failures why does it have to deal with unrelated out-of-order arguments as well?

Reply:
Here it is deciding whether or not to integrate the diagnosis. 
It is different from the logic in the diagnosis.
It deals with `OutOfOrder` because diagnostics about it may also be aggregated.
Please read my reply about `findLabelingFixes` for more information.